### ISSUE TITLE
ASN Lookup on RFC1918 IPs fails

### ISSUE THREAD

**Issue by philhagen** [logstash-plugins/logstash-filter-geoip#30]
Found that doing an GeoIP lookup on a private IP against the ASN database causes Logstash to hang, requiring a SIGKILL.

`# cat /etc/centos-release`
`CentOS Linux release 7.1.1503 (Core)`

`# /opt/logstash/bin/logstash -V`
`logstash 1.5.0`

GeoIP database is latest as of this filing
`# md5sum  /usr/local/share/GeoIP/GeoIPASNum.dat`
`d5672dde5cd81286aaaf87ab7defc690  /usr/local/share/GeoIP/GeoIPASNum.dat`

Configuration (distilled down from production configuration)
https://gist.github.com/philhagen/5f6d1df7fb32a958d523

Input samples (only differences between these are the actual SRC and DST IP addresses - embedded ICMP message fields are not parsed):
`Jun  8 11:23:34 fw_ids kernel: FW DENY_OUTPUT: IN= OUT=eth0 SRC=10.3.16.99 DST=10.3.59.52 LEN=265 TOS=0x00 PREC=0xC0 TTL=64 ID=41078 PROTO=ICMP TYPE=3 CODE=1 [SRC=10.3.59.52 DST=10.3.58.204 LEN=237 TOS=0x00 PREC=0x00 TTL=127 ID=180 DF PROTO=UDP SPT=52091 DPT=514 LEN=217 ]`

`Jun  8 11:23:34 fw_ids kernel: FW DENY_OUTPUT: IN= OUT=eth0 SRC=192.168.15.109 DST=8.8.8.8 LEN=265 TOS=0x00 PREC=0xC0 TTL=64 ID=41078 PROTO=ICMP TYPE=3 CODE=1 [SRC=10.3.59.52 DST=10.3.58.204 LEN=237 TOS=0x00 PREC=0x00 TTL=127 ID=180 DF PROTO=UDP SPT=52091 DPT=514 LEN=217 ]`

`Jun  8 11:23:34 fw_ids kernel: FW DENY_OUTPUT: IN= OUT=eth0 SRC=1.2.3.4 DST=172.17.51.10 LEN=265 TOS=0x00 PREC=0xC0 TTL=64 ID=41078 PROTO=ICMP TYPE=3 CODE=1 [SRC=10.3.59.52 DST=10.3.58.204 LEN=237 TOS=0x00 PREC=0x00 TTL=127 ID=180 DF PROTO=UDP SPT=52091 DPT=514 LEN=217 ]`

`Jun  8 11:23:34 fw_ids kernel: FW DENY_OUTPUT: IN= OUT=eth0 SRC=1.2.3.4 DST=8.8.8.8 LEN=265 TOS=0x00 PREC=0xC0 TTL=64 ID=41078 PROTO=ICMP TYPE=3 CODE=1 [SRC=10.3.59.52 DST=10.3.58.204 LEN=237 TOS=0x00 PREC=0x00 TTL=127 ID=180 DF PROTO=UDP SPT=52091 DPT=514 LEN=217 ]`

Results:
Samples 1-3 result in the following:
```javascript
Exception in filterworker {"exception"=>#<NoMethodError: undefined method `to_hash' for "*AS18144 Energia Communications,Inc.":String>, "backtrace"=>["/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-filter-geoip-0.1.9/lib/logstash/filters/geoip.rb:119:in `filter'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/filters/base.rb:162:in `multi_filter'", "org/jruby/RubyArray.java:1613:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/filters/base.rb:159:in `multi_filter'", "(eval):1468:in `cond_func_65'", "org/jruby/RubyArray.java:1613:in `each'", "(eval):1465:in `cond_func_65'", "(eval):482:in `filter_func'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/pipeline.rb:221:in `filterworker'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/logstash-core-1.5.0-java/lib/logstash/pipeline.rb:157:in `start_filters'"], :level=>:error}
```

When the filter fails, Logstash cannot be stopped with a SIGTERM - only a SIGKILL.

Sample 4 (which contains two non-RFC1918 IPs) parses fine:
```json5
{
        "message" => "Jun  8 11:23:34 fw_ids kernel: FW DENY_OUTPUT: IN= OUT=eth0 SRC=1.2.3.4 DST=8.8.8.8 LEN=265 TOS=0x00 PREC=0xC0 TTL=64 ID=41078 PROTO=ICMP TYPE=3 CODE=1 [SRC=10.3.59.52 DST=10.3.58.204 LEN=237 TOS=0x00 PREC=0x00 TTL=127 ID=180 DF PROTO=UDP SPT=52091 DPT=514 LEN=217 ]",
       "@version" => "1",
     "@timestamp" => "2015-06-12T20:40:21.165Z",
           "type" => "archive-syslog",
           "host" => "logstash2.identityvector.com",
            "ips" => [
        [0] "1.2.3.4",
        [1] "8.8.8.8"
    ],
      "ipt_label" => "Jun  8 11:23:34 fw_ids kernel: FW DENY_OUTPUT",
     "ipt_outdev" => "eth0",
      "ipt_srcip" => "1.2.3.4",
      "ipt_dstip" => "8.8.8.8",
        "ipt_ttl" => "127",
      "ipt_proto" => "UDP",
    "ipt_srcport" => "52091",
    "ipt_dstport" => "514",
           "tags" => [
        [0] "got_iptables",
        [1] "parse_done"
    ],
     "ipt_srcgeo" => {
                  "number" => "AS15169",
                     "asn" => "Google Inc.",
           "country_code2" => "US",
           "country_code3" => "USA",
            "country_name" => "United States",
          "continent_code" => "NA",
             "region_name" => "WA",
               "city_name" => "Mukilteo",
             "postal_code" => "98275",
                "latitude" => 47.91300000000001,
               "longitude" => -122.30420000000001,
                "dma_code" => 819,
               "area_code" => 425,
                "timezone" => "America/Los_Angeles",
        "real_region_name" => "Washington",
                "location" => [
            [0] -122.30420000000001,
            [1] 47.91300000000001
        ]
    },
     "ipt_dstgeo" => {
                  "number" => "AS15169",
                     "asn" => "Google Inc.",
           "country_code2" => "US",
           "country_code3" => "USA",
            "country_name" => "United States",
          "continent_code" => "NA",
             "region_name" => "CA",
               "city_name" => "Mountain View",
             "postal_code" => "94040",
                "latitude" => 37.385999999999996,
               "longitude" => -122.0838,
                "dma_code" => 807,
               "area_code" => 650,
                "timezone" => "America/Los_Angeles",
        "real_region_name" => "California",
                "location" => [
            [0] -122.0838,
            [1] 37.385999999999996
        ]
    }
}
```

It should also be noted that the same IP lookups work fine against the latest GeoLiteCity database - only the ASNs fail.

**Comment by philhagen** [comment]
This may be related to Issue #25.

**Comment by philhagen** [comment]
Well that was silly of me.  Sorry to waste time - this is all fixed with version 0.1.10 of the plugin, in #24.
Carry on - hope filing this helps someone else avoid the headache someday...
